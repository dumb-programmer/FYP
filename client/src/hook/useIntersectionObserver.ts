import { useEffect, useRef, useState } from "react";

export default function useIntersectionObserver() {
  const ref = useRef<HTMLDivElement>(null);
  const [inView, setInView] = useState(false);

  useEffect(() => {
    const options = {
      root: null,
      rootMargin: "0px",
      threshold: 1.0,
    };
    const observer = new IntersectionObserver((entries) => {
      entries.forEach((entry) => {
        if (entry.target === ref.current) {
          setInView(entry.isIntersecting);
        }
      });
    }, options);

    if (ref.current) {
      observer.observe(ref.current);
    }

    return () => {
      observer.disconnect();
    };
  }, []);

  return { ref, inView };
}
