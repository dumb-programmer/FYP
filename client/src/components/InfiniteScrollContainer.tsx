import useIntersectionObserver from "@/hook/useIntersectionObserver";
import { useEffect, useRef } from "react";

interface InfiniteScrollContainerProps extends React.HTMLProps<HTMLDivElement> {
    hasNextPage: boolean | undefined;
    isFetchingNextPage: boolean | undefined;
    fetchNextPage: () => void;
    children: React.ReactNode;
}

export default function InfiniteScrollContainer({ hasNextPage, isFetchingNextPage, fetchNextPage, children, ...props }: InfiniteScrollContainerProps) {
    const containerRef = useRef<HTMLDivElement>(null);
    const { inView, ref } = useIntersectionObserver();

    useEffect(() => {
        if (inView && !isFetchingNextPage && hasNextPage) {
            fetchNextPage();
            containerRef?.current?.scrollBy({
                top: 20,
                behavior: "smooth",
            });
        }
    }, [inView]);

    return <div ref={containerRef} {...props}>
        <div ref={ref}></div>
        {
            children
        }
    </div>
}