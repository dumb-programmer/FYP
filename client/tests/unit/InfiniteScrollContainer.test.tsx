import React from "react";
import { render } from "@testing-library/react";
import { expect, describe, it, vi, afterEach, beforeEach } from "vitest";
import InfiniteScrollContainer from "../../src/components/InfiniteScrollContainer";
import useIntersectionObserver from "../../src//hooks/useIntersectionObserver";

vi.mock("../../src/hooks/useIntersectionObserver", () => ({ default: vi.fn(() => ({ inView: true, ref: vi.fn() })) }))

const mockFetchNextPage = vi.fn();
const renderComponent = (hasNextPage: boolean, isFetchingNextPage: boolean) => render(<InfiniteScrollContainer hasNextPage={hasNextPage} isFetchingNextPage={isFetchingNextPage} fetchNextPage={mockFetchNextPage}><div className="w-96"></div></InfiniteScrollContainer>);

describe("InfiniteScrollContainer", () => {

    beforeEach(() => {
        HTMLDivElement.prototype.scrollBy = vi.fn();
    })

    afterEach(() => {
        vi.clearAllMocks();
    });

    it("renders correctly", () => {
        const { container } = renderComponent(true, true);

        expect(container).toMatchSnapshot();
    });

    it("fetchNextPage is called when inView is true, and isFetchingNextPage is false, and hasNextPage is true", () => {
        renderComponent(true, false);

        expect(mockFetchNextPage).toHaveBeenCalled();

        expect(HTMLDivElement.prototype.scrollBy).toHaveBeenCalled();
        expect(HTMLDivElement.prototype.scrollBy).toHaveBeenCalledWith({ top: 20, behavior: "smooth" });
    });

    it("fetchNextPage is not called when inView is true, and isFetchingNextPage is true, and hasNextPage is true", () => {
        renderComponent(true, true);

        expect(mockFetchNextPage).not.toHaveBeenCalled();
        expect(HTMLDivElement.prototype.scrollBy).not.toHaveBeenCalled();
    });

    it("fetchNextPage is not called when inView is true, and isFetchingNextPage is false, and hasNextPage is false", () => {
        renderComponent(false, false);

        expect(mockFetchNextPage).not.toHaveBeenCalled();
        expect(HTMLDivElement.prototype.scrollBy).not.toHaveBeenCalled();
    });

    it("fetchNextPage is not called when inView is false", () => {
        useIntersectionObserver.mockReturnValue({
            inView: false,
            ref: vi.fn(),
        });

        renderComponent(false, false);

        expect(mockFetchNextPage).not.toHaveBeenCalled();
        expect(HTMLDivElement.prototype.scrollBy).not.toHaveBeenCalled();
    });
})