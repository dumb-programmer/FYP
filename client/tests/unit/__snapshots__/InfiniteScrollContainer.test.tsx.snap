// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`InfiniteScrollContainer > renders correctly 1`] = `
<div>
  <div>
    <div />
    <div
      class="w-96"
    />
  </div>
</div>
`;
